Set profile picture more easily, remove old, broken implementation

marcusxss opened the pull request:
Adds functionality to more easily set profile picture, based on my issue #14984 

You hover over your profile picture on any of your profile pages, and this appears in...

...Frio:
![image](https://github.com/user-attachments/assets/6a55e5e8-79af-4adb-baba-f429a5088dde)

...Vier:
![image](https://github.com/user-attachments/assets/8c91b2d5-0133-4581-af2e-dd940264249b)

...and it takes you to `/profile/<username/photos`, where an existing profile picture can be chosen or a new one uploaded like normally:
![image](https://github.com/user-attachments/assets/a751de68-5046-407a-a0ce-d88c0e5797b5)

Also I removed the older, broken, somewhat similar feature, which I only found out about while making this.
It wasn't working, because apparently this logic, meant to differentiate if you're seeing the user as yourself or using the "View as" functionality, is apparently always set:
```php
!isset($profile['self'])
```
...and so this edit box never showed up on hover:
![image](https://github.com/user-attachments/assets/eb1fe1ee-bd38-44fe-8582-ec628df3824e)

I wasn't quite sure whether to do anything with the equivalent template that isn't under one of the themes (`/templates/profile/vcard.tpl`).
Is that one only used by themes that don't have their own `vcard.tpl`, so they work as overrides?

Also I could remove the `action` css class now it's no longer used by frio and vier,  except the `quattro` theme still uses it, so I don't know?

Also, a minor change, I centered the user text under the profile picture in Frio, like Bookface.

A limitation of this PR is that that "change profile picture" box ALSO appears when you're using "View as".
I wanted to not show it in that case, but I'm not sure how to check whether "View as" is active. Clearly the previously used approach was no longer working.

Also, since profile images are rounded in bookface unlike frio and vier currently, this doesn't look too good in bookface, but bookface should just be able to override that for the next release?

marcusxss committed:
Change profile picture more easily, remove old implementation

marcusxss commented:
Hm, seems there's an issue with the final changes I made.
Investigating.

Edit: Oh, it was because a variable would potentially not be defined. Now that's no longer the case.

annando commented:
@mf-fx we now created a RC branch. Please have a look at your open PRs and decide which of them should be part of the upcoming release and possibly change the target branch.

marcusxss commented:
@annando I've rebased my changes on the new target branch and updated the target branch on the PR.

Do you know a way to check whether "view as" is active, so I could not show this option to change profile picture when that is active?

annando commented:
See here: https://github.com/friendica/friendica/blob/2025.07-rc/src/Module/Profile/Profile.php#L149
The variable `$view_as_contact_id` will be set to a non zero value in this case.

marcusxss committed:
Change profile picture more easily: Detect if 'View as' is active

marcusxss commented:
Thanks! Locally it doesn't show when "view as" is active now.

annando commented on src/Module/Profile/Profile.php:
```diff
@@ -116,7 +116,7 @@ protected function rawContent(array $request = [])
 	protected function content(array $request = []): string
 	{
-		$profile = ProfileModel::load($this->appHelper, $this->parameters['nickname'] ?? '');
+		$profile = ProfileModel::load($this->appHelper, $this->parameters['nickname'] ?? '', true, $request['viewas'] == 0);
```
You have to check if `$request['viewas']` is set at all. Also when `$request['viewas']` is set to a non zero value, then you are in the "view as" mode.

marcusxss replied:
You're right. Inverted the variable name + changed it to assume `view as` is inactive if it isn't set, with `??`.

Does it look okay now?

annando commented:
What about a different solution? You could move the lines 145 to 172 (the block, where the `view_as` variables are set) to the beginning of the function. Then you could pass the variable `$view_as_contact_id` to the `load` function.

marcusxss commented:
I did consider it, but I need to pass it to the function which returns the profile object, but that `view as` block below assumes that the `profile` object has already been returned, fx. here:
```php
$is_owner = $this->session->getLocalUserId() == $profile['uid'];
$o        = self::getTabsHTML('profile', $is_owner, $profile['nickname'], $profile['hide-friends']);
```
...that `is_owner` variable is checked before `view_as_contact_id` is set:
```php
if ($is_owner) {
	$view_as_contact_id = intval($request['viewas'] ?? 0);
```
...?

marcusxss commented:
Of course, if you have very specific ideas for a rewrite feel free to do so.
You know PHP and Friendica and the best practices with them much better than I do.

annando commented:
See this suggestion:
```
	protected function content(array $request = []): string
	{
		$owner = User::getByNickname($this->parameters['nickname'] ?? '', ['uid']);
		if (!$owner) {
			throw new HTTPException\NotFoundException($this->t('Profile not found.'));
		}

		$is_owner              = $this->session->getLocalUserId() == $owner['uid'];
		$view_as_contacts      = [];
		$view_as_contact_id    = 0;
		$view_as_contact_alert = '';
		if ($is_owner) {
			$view_as_contact_id = intval($request['viewas'] ?? 0);

			$view_as_contacts = Contact::selectToArray(['id', 'name'], [
				'uid'     => $this->session->getLocalUserId(),
				'rel'     => [Contact::FOLLOWER, Contact::SHARING, Contact::FRIEND],
				'network' => Protocol::DFRN,
				'blocked' => false,
			]);

			$view_as_contact_ids = array_column($view_as_contacts, 'id');

			// User manually provided a contact ID they aren't privy to, silently defaulting to their own view
			if (!in_array($view_as_contact_id, $view_as_contact_ids)) {
				$view_as_contact_id = 0;
			}

			if (($key = array_search($view_as_contact_id, $view_as_contact_ids)) !== false) {
				$view_as_contact_alert = $this->t(
					'You\'re currently viewing your profile as <b>%s</b> <a href="%s" class="btn btn-sm pull-right">Cancel</a>',
					htmlentities($view_as_contacts[$key]['name'], ENT_COMPAT, 'UTF-8'),
					'profile/' . $this->parameters['nickname'] . '/profile'
				);
			}
		}

		$profile = ProfileModel::load($this->appHelper, $this->parameters['nickname'], true, $view_as_contact_id);
		if (!$profile) {
			throw new HTTPException\NotFoundException($this->t('Profile not found.'));
		}

		$remote_contact_id = $this->session->getRemoteContactID($profile['uid']);

		if ($this->config->get('system', 'block_public') && !$this->session->isAuthenticated()) {
			return Login::form();
		}

		if (!empty($profile['hidewall']) && !$this->session->isAuthenticated()) {
			$this->baseUrl->redirect('profile/' . $profile['nickname'] . '/restricted');
		}

		if (!empty($profile['page-flags']) && in_array($profile['page-flags'], [User::PAGE_FLAGS_COMMUNITY, User::PAGE_FLAGS_COMM_MAN])) {
			$this->page['htmlhead'] .= '<meta name="friendica.community" content="true" />' . "\n";
		}

		$this->page['htmlhead'] .= $this->buildHtmlHead($profile, $this->parameters['nickname']);

		Nav::setSelected('home');

		$o        = self::getTabsHTML('profile', $is_owner, $profile['nickname'], $profile['hide-friends']);

		$basic_fields = [];

		$basic_fields += self::buildField('fullname', $this->t('Full Name:'), $this->cleanInput($profile['uri-id'], $profile['name']));

		if (Feature::isEnabled($profile['uid'], Feature::MEMBER_SINCE)) {
			$basic_fields += self::buildField(
				'membersince',
				$this->t('Member since:'),
				DateTimeFormat::local($profile['register_date'])
			);
		}

		if (!empty($profile['dob']) && $profile['dob'] > DBA::NULL_DATE) {
			$year_bd_format  = $this->t('j F, Y');
			$short_bd_format = $this->t('j F');

			$dob = $this->l10n->getDay(
				intval($profile['dob']) ?
					DateTimeFormat::utc($profile['dob'] . ' 00:00 +00:00', $year_bd_format)
					: DateTimeFormat::utc('2001-' . substr($profile['dob'], 5) . ' 00:00 +00:00', $short_bd_format)
			);

			$basic_fields += self::buildField('dob', $this->t('Birthday:'), $dob);

			if ($age = Temporal::getAgeByTimezone($profile['dob'], $profile['timezone'])) {
				$basic_fields += self::buildField('age', $this->t('Age: '), $this->tt('%d year old', '%d years old', $age));
			}
		}
```

marcusxss commented:
Thanks! 
I've attempted to copy in your rewrite as-is, and adjusted the other functions to now take an `int` instead of a `bool`.
Functionally it appears to remain working, so the photos URL and profile picture change box doesn't appear when "View as" is active.

haheute commented:
I really don't like that the username and user info text are now centered. If there is any chance, please undo it. (or maybe only center it for mobile narrow screens)
